### PR TITLE
Implement NLUMessageConverter

### PR DESCRIPTION
**Pull request by ancalita**
**Proposed changes**:
- implements #9326 

**Status (please check what you already did)**:
- [X] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/main/changelog) for instructions)
- [X] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

File: rasa/graph_components/converters/nlu_message_converter.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        config: Dict[Text, Any],
+        model_storage: ModelStorage,
+        resource: Resource,
+        execution_context: ExecutionContext,
+    ) -> NLUMessageConverter:
+        """Creates component (see parent class for full docstring)."""
+        return cls(config)
+
+    @staticmethod
+    def convert_user_message(message: Optional[UserMessage]) -> List[Optional[Message]]:
```
Comment (wochinge) [inline rasa/graph_components/converters/nlu_message_converter.py:44]:
```suggestion
    def convert_user_message(message: Optional[UserMessage]) -> List[Message]:
```

File: rasa/graph_components/converters/nlu_message_converter.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        """Converts user message into Message object.
+
+        Returns:
+            List containing only one instance of Message.
+            Else empty list if user message is None.
+        """
+        if message:
+            data = dict()
```
Comment (wochinge) [inline rasa/graph_components/converters/nlu_message_converter.py:52]: optional: maybe something like this is slightly easier to read?

```python
data = {
  TEXT: message.text,
  "input_channel": message.input_channel,
  "message_id: message.message_id,
  ...
}
```

File: rasa/graph_components/converters/nlu_message_converter.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+            data[TEXT] = message.text
+            data["input_channel"] = message.input_channel
```
Comment (wochinge) [inline rasa/graph_components/converters/nlu_message_converter.py:54]: Do we actually need the input channel here? [This code](https://github.com/RasaHQ/rasa/blob/f30fade97c9fd50b7156aa7d8513bc2fc3527a78/rasa/core/processor.py#L589-L591) is currently not doing that. From my intuition it's fine if the [`NLUPredictionToHistoryAdder`](https://github.com/RasaHQ/rasa/issues/9327) takes care of passing the `input_channel` to the `UserUttered` event.

File: rasa/graph_components/converters/nlu_message_converter.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+            data["message_id"] = message.message_id
+            data["metadata"] = message.metadata
+
+            if message.parse_data:
```
Comment (wochinge) [inline rasa/graph_components/converters/nlu_message_converter.py:58]: There should never be a `message.parse_data` at this point as the `NLUMessageConverter` comes before the graph section which adds tokens, features, intent / entity predictions. So you always should just have a plain `UserMessage` object which you need to convert to `Message` object which then passes through what used to be the `NLU pipeline"

File: rasa/graph_components/converters/nlu_message_converter.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    def get_default_config() -> Dict[Text, Any]:
+        """Returns default configuration (see parent class for full docstring)."""
+        return {
+            "remove_duplicates": True,
+            "unique_last_num_states": None,
+            "augmentation_factor": 50,
+            "tracker_limit": None,
+            "use_story_concatenation": True,
+            "debug_plots": False,
+        }
```
[the remark this answers is not in the document]
Comment (ancalita) [reply rasa/graph_components/converters/nlu_message_converter.py:30]: I wasn't actually sure what should be the default config for this component so I was hoping to find out during the review 😄 So it's actually an empty `{ }` and therefore fine to delete these methods?

File: rasa/graph_components/converters/nlu_message_converter.py (hunk 1 above)
Comment (ancalita) [reply rasa/graph_components/converters/nlu_message_converter.py:44]: if this method can return an empty list, why are we removing the `Optional` part? 🤔 Or does `Optional` allow for list of `None` and that's why it should be removed?

File: rasa/graph_components/converters/nlu_message_converter.py (hunk 4 above)
Comment (ancalita) [reply rasa/graph_components/converters/nlu_message_converter.py:58]: Good point! I got confused by the fact that `parse_data` is an existing attribute of `UserMessage` - why is this even present if this is at the start of the NLU pipeline?

File: rasa/graph_components/converters/nlu_message_converter.py (hunk 3 above)
Comment (ancalita) [reply rasa/graph_components/converters/nlu_message_converter.py:54]: I wasn't sure how to check what does `Message.data` contain to be sure if `input_channel` is relevant or not 🤔

File: rasa/graph_components/converters/nlu_message_converter.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+        """Converts user message into Message object.
+
+        Returns:
+            List containing only one instance of Message.
+            Else empty list if user message is None.
+        """
+        if message:
+            data = {
+                TEXT: message.text,
+                "input_channel": message.input_channel,
```
Comment (wochinge) [inline rasa/graph_components/converters/nlu_message_converter.py:38]: this one can be dropped, no?

File: tests/test_graph_components/converters/test_nlu_message_converter.py [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+
+    if text:
+        message = UserMessage(text=text)
+        nlu_message = component.convert_user_message(message)
+        assert len(nlu_message) == 1
+        assert isinstance(nlu_message[0], Message)
+
+        assert nlu_message[0].get("text") == "Hi"
+        assert nlu_message[0].get("metadata") is None
+    else:
```
Comment (wochinge) [inline tests/test_graph_components/converters/test_nlu_message_converter.py:36]: optional: This is in my opinion a separate unit test as it tests for a different aspect (e.g. https://blog.colinbreck.com/in-defence-of-testing-one-thing-per-test-method/)

File: rasa/graph_components/converters/nlu_message_converter.py (hunk 4 above)
Comment (wochinge) [reply rasa/graph_components/converters/nlu_message_converter.py:43]: seems to be a lazy hack which is only used by this code https://github.com/RasaHQ/rasa/blob/a1afc66ee482955f89bfc4318b4cd3907f8e2378/rasa/server.py#L1038

File: rasa/graph_components/converters/nlu_message_converter.py (hunk 3 above)
Comment (wochinge) [reply rasa/graph_components/converters/nlu_message_converter.py]: I'd check the line in the existing code where `Message` is created. There it doesn't add the `input_channel` to `Message`)

File: tests/test_graph_components/converters/test_nlu_message_converter.py [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+    )
+
+    if text:
+        message = UserMessage(text=text)
+        nlu_message = component.convert_user_message(message)
+        assert len(nlu_message) == 1
+        assert isinstance(nlu_message[0], Message)
+
+        assert nlu_message[0].get("text") == "Hi"
+        assert nlu_message[0].get("metadata") is None
```
Comment (wochinge) [inline tests/test_graph_components/converters/test_nlu_message_converter.py:35]: can we please also have a test with metadata? I know that some community members use that to pass extra information into their NLU components

Comment (ancalita) [reply tests/test_graph_components/converters/test_nlu_message_converter.py:35]: sure, what kind of key-value info goes into metadata?

Comment (wochinge) [reply tests/test_graph_components/converters/test_nlu_message_converter.py:35]: any 🙈 That's completely up to the bot developer. So metadata would be just random dict basically.

File: rasa/graph_components/converters/nlu_message_converter.py (hunk 3 above)
Comment (ancalita) [reply rasa/graph_components/converters/nlu_message_converter.py]: > I'd check the line in the existing code where Message is created. 

Are you referring to this [line](https://github.com/RasaHQ/rasa/blob/7f0bb29b1a40e237931f6f431813e08654bea1fb/rasa/nlu/model.py#L471)? 
Ok, yup, will drop the `input_channel`.

Comment (wochinge) [reply rasa/graph_components/converters/nlu_message_converter.py]: yep exactly 👍🏻

**Review by wochinge — APPROVED** [review]
💯

**Pull request merged**